src/shims/unix/foreign_items.rs

```diff
@@ -440,13 +440,34 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             // Miscellaneous
             "isatty" => {
                 let [fd] = this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
-                this.read_scalar(fd)?.to_i32()?;
-                // "returns 1 if fd is an open file descriptor referring to a terminal; otherwise 0 is returned, and errno is set to indicate the error"
-                // FIXME: we just say nothing is a terminal.
-                let enotty = this.eval_libc("ENOTTY")?;
-                this.set_last_error(enotty)?;
-                this.write_null(dest)?;
-            }
+                #[cfg_attr(not(unix), allow(unused))]
+                let fd = this.read_scalar(fd)?.to_i32()?;
+                #[cfg(unix)]
+                {
+                    // "returns 1 if fd is an open file descriptor referring to a terminal;
+                    // otherwise 0 is returned, and errno is set to indicate the error"
+                    let is_tty = unsafe {
+                        libc::isatty(fd)
+                    };
+                    if is_tty == 1 {
+                        this.write_scalar(Scalar::from_i32(1), dest)?;
+                    } else {
+                        let enotty = this.eval_libc("ENOTTY")?;
+                        let errno = std::io::Error::last_os_error().raw_os_error().map_or(enotty, Scalar::from_i32);
+                        this.set_last_error(errno)?;
+                        this.write_null(dest)?;
```

Comment:
Please no interesting logic in this file, it should be mostly dispatching.  This should go into some other file in src/shims/unix/ or a new file if none of the existing one fits.